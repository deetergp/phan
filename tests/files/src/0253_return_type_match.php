<?php
class C253_1 {
    /** @return string */
    public function f1() : int {
        return 7;
    }
    /** @return string[] */
    public function f2() : array {
        return array('a', 'b', 'c');
    }
    /** @return $this */
    public function f3() : C253_1 {
        return $this;
    }
    /** @return static */
    public static function f4() : C253_1 {
        return new static();
    }

    /** @return iterable (Widening) */
    public function a5() : array {
        return array('a', 'b', 'c');
    }
    /** @return array (narrowing, allowed) */
    public function a6() : iterable {
        return array('a', 'b', 'c');
    }
    /** @return Traversable (narrowing, allowed) */
    public function a7() : iterable {
        yield 'a';
    }
    /** @return string (incompatible, not allowed) */
    public function a8() : iterable {
        return array('a', 'b', 'c');
    }
    /** @return static (narrowing) */
    public function a9() : self {
        return new static();
    }

    /** @return ?int (widening, not allowed) */
    public function a10() : int {
        return 42;
    }

    /** @return int (phpdoc narrowing, currently assumed to be a mistake) */
    public function a11() : ?int {
        return 42;
    }
}
class C253_2 extends C253_1 {
    /** @return C253_2 (narrowing, allowed) */
    public function f5() : C253_1 {
        return new C253_2;
    }
    /** @return C253_1 (widening, not allowed) */
    public function f6() : C253_2 {
        return new C253_2;
    }
    /** @return C253_1 (widening, not allowed) */
    public function f7() : self {
        return new static();
    }
}
