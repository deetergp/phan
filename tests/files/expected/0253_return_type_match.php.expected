%s:4 PhanTypeMismatchDeclaredReturn Doc-block of f1 contains declared return type string which is incompatible with the return type int declared in the signature
%s:21 PhanTypeMismatchDeclaredReturn Doc-block of a5 contains declared return type iterable which is incompatible with the return type array declared in the signature
%s:33 PhanTypeMismatchDeclaredReturn Doc-block of a8 contains declared return type string which is incompatible with the return type iterable declared in the signature
%s:42 PhanTypeMismatchDeclaredReturn Doc-block of a10 contains declared return type ?int which is incompatible with the return type int declared in the signature
%s:47 PhanTypeMismatchDeclaredReturnNullable Doc-block of a11 has declared return type int which is not a permitted replacement of the nullable return type ?int declared in the signature ('?T' should be documented as 'T|null' or '?T')
%s:57 PhanTypeMismatchDeclaredReturn Doc-block of f6 contains declared return type \C253_1 which is incompatible with the return type \C253_2 declared in the signature
%s:61 PhanTypeMismatchDeclaredReturn Doc-block of f7 contains declared return type \C253_1 which is incompatible with the return type \C253_2 declared in the signature
